src/main/java/frc/trigon/robot/poseestimation/robotposesources/RobotPoseSource.java

```diff
@@ -76,11 +67,57 @@ public double getLastResultTimestamp() {
         return inputs.lastResultTimestamp;
     }
 
-    private Pose2d getUnCachedRobotPose() {
-        final Pose3d cameraPose = doubleArrayToPose3d(inputs.cameraPose);
+    public Matrix<N3, N1> calculateStdDevs() {
+        if (inputs.distanceFromBestTag < RobotPoseSourceConstants.MAXIMUM_DISTANCE_FROM_TAG_FOR_PNP_METERS)
+            return solvePNPAverageDistanceToStdDevs();
+        return assumedHeadingAverageDistanceToStdDevs();
+    }
+
+    /**
+     * If the robot is close enough to the tag, it gets the estimated solve PNP pose.
+     * If it's too far, it assumes the robot's heading and calculates its position from there.
+     *
+     * @param distanceFromBestTag the average of the distance from the best visible tag
+     * @return the robot's pose
+     */
+    private Pose2d calculateBestRobotPose(double distanceFromBestTag) {
+        if (distanceFromBestTag < RobotPoseSourceConstants.MAXIMUM_DISTANCE_FROM_TAG_FOR_PNP_METERS)
+            return cameraPoseToRobotPose(inputs.solvePNPPose);
+        return calculateAssumedRobotHeadingPose().toPose2d();
+    }
+
+    /**
+     * Calculates the robot's pose by assuming its heading.
+     *
```

Comment:
This assumes one more thing
Talk about how this is more robust

Reply:
What else does it assume?

Reply:
Read through the code